libdevcrypto/CryptoPP.cpp

```diff
@@ -31,6 +32,124 @@ static_assert(dev::Secret::size == 32, "Secret key must be 32 bytes.");
 static_assert(dev::Public::size == 64, "Public key must be 64 bytes.");
 static_assert(dev::Signature::size == 65, "Signature must be 65 bytes.");
 
+bytes Secp256k1::eciesKDF(Secret _z, bytes _s1, unsigned kdByteLen)
+{
+	// interop w/go ecies implementation
+	
+	// for sha3, blocksize is 136 bytes
+	// for sha256, blocksize is 64 bytes
+	auto reps = ((kdByteLen + 7) * 8) / (64 * 8);
+	bytes ctr({0, 0, 0, 1});
+	bytes k;
+	CryptoPP::SHA256 ctx;
+	for (unsigned i = 0; i <= reps; i++)
+	{
+		ctx.Update(ctr.data(), ctr.size());
+		ctx.Update(_z.data(), Secret::size);
+		ctx.Update(_s1.data(), _s1.size());
+		// append hash to k
+		bytes digest(32);
+		ctx.Final(digest.data());
+		ctx.Restart();
+		
+		k.reserve(k.size() + h256::size);
+		move(digest.begin(), digest.end(), back_inserter(k));
+		
+		if (ctr[3]++ && ctr[3] != 0) {
+			continue;
+		} else if (ctr[2]++ && ctr[2] != 0) {
+			continue;
+		} else if (ctr[1]++ && ctr[1] != 0) {
+			continue;
+		} else
+			ctr[0]++;
+	}
+	
+	k.resize(kdByteLen);
+	return move(k);
+}
+
+void Secp256k1::encryptECIES(Public const& _k, bytes& io_cipher)
+{
+	// interop w/go ecies implementation
+	auto r = KeyPair::create();
+	h256 z;
+	ecdh::agree(r.sec(), _k, z);
+	auto key = eciesKDF(z, bytes(), 512);
+	bytesConstRef eKey = bytesConstRef(&key).cropped(0, 32);
+	bytesRef mKey = bytesRef(&key).cropped(32, 32);
+	CryptoPP::SHA256 ctx;
+	ctx.Update(mKey.data(), mKey.size());
+	ctx.Final(mKey.data());
+	
+	bytes cipherText;
+	encryptSymNoAuth(*(Secret*)eKey.data(), bytesConstRef(&io_cipher), cipherText, h128());
```

Comment:
Petition for FixedHashConstRef, h256ConstRef, and h128ConstRef, h256Ref, h128Ref, etc?
Style: Just not going to copy key material. Also: encryptSymNoAuth will likely take h128 and h256 instead of Secret. Still will use *(h128*) so as to not copy key material.